### PR TITLE
In `to_thread_run_sync()`, add `abandon_on_cancel=` as an alias for the `cancellable=` flag

### PR DESCRIPTION
**Pull request by richardsheridan**
As suggested in https://github.com/python-trio/trio/pull/2392#discussion_r1354105279

### CODE REVIEW

File: newsfragments/2841.feature.rst
Comment (A5rocks) [inline newsfragments/2841.feature.rst:3]: Hmm it's not exactly deprecated. I'm not sure how to best phase "replaced temporary kwarg name" or whatever though. I feel like it's important information that `cancellable` was never released.

Comment (richardsheridan) [reply newsfragments/2841.feature.rst:3]: I'm happy to simply not say deprecated anywhere!

Comment (A5rocks) [reply newsfragments/2841.feature.rst:3]: Yeah that probably works

Comment (A5rocks) [reply newsfragments/2841.feature.rst]: Oh wait gahhh I shouldn't do this while sleepy, I only saw the changes that replaced the kwarg, didn't realize you kept it for compatibility 😭 

Sorry... if we are going to keep the kwarg, then deprecation is good. I thought the function that "cancelled" is on was added in this release meaning we could just replace the name without deprecation period.

Comment (richardsheridan) [reply newsfragments/2841.feature.rst]: Haha no worries, but I won't be able to revert it today. Feel free to merge either version in the mean while.

**Review by TeamSpen210 — APPROVED** [review]
That's a much more understandable name. It's not necessary, but we could use `@overload` to have type checkers prevent you from passing both parameters:
```py
@overload
async def to_thread_run_sync(
    sync_fn: Callable[..., RetT],
    *args: object,
    thread_name: str | None = None,
    abandon_on_cancel: bool,
    limiter: CapacityLimiter | None = None,
) -> RetT: ...
@overload
async def to_thread_run_sync(
    sync_fn: Callable[..., RetT],
    *args: object,
    thread_name: str | None = None,
    cancellable: bool = False,
    limiter: CapacityLimiter | None = None,
) -> RetT: ...
```

**Review by A5rocks — APPROVED** [review]

**Review by jakkdl — CHANGES_REQUESTED** [review]
If `cancellable` is actually deprecated, as it's phrased in the newsfragment, then it should raise a `DeprecationWarning` if `cancellable` is specified.
Alternatives would be to forever support both arguments (yuck), or to make it deprecated later for some reason.

**Comment by richardsheridan** [comment]
I'd sooner abandon this pull request than start raising warnings everywhere. The extra maintenance burden is 3 easy lines of code on our end, but a warning will send our users scrambling to perform busywork. Someone else should add a warning in a future PR for a real reason, like `cancellable will be removed in 1.0`, at the time such a decision is made to balance user and maintainer burdens.

File: newsfragments/2841.feature.rst
Comment (jakkdl) [inline newsfragments/2841.feature.rst:3]:
```suggestion
To better reflect the underlying thread handling semantics,
`trio.to_thread.run_sync` gained a new keyword argument ``abandon_on_cancel``
that can be used instead of the older ``cancellable`` flag (that might get fully deprecated in the future).
```

File: docs/source/reference-core.rst
Comment (oremanj) [inline docs/source/reference-core.rst:1839]: I think we can remove this parenthetical, it no longer makes sense since the name isn't confusing anymore

File: newsfragments/2841.feature.rst
Comment (oremanj) [inline newsfragments/2841.feature.rst:2]: This newsfragment should get a `.removal.` name (not `.feature.`) so it shows up in the "Deprecations and removals" section of the release notes.

I would write this as
```
To better reflect the underlying thread handling semantics,
the keyword argument for `trio.to_thread.run_sync` that was
previously called ``cancellable`` is now named ``abandon_on_cancel``.
It still does the same thing -- allow the thread to be abandoned
if the call to `trio.to_thread.run_sync` is cancelled -- but since we now
have other ways to propagate a cancellation without abandoning
the thread, "cancellable" has become somewhat of a misnomer.
The old ``cancellable`` name is now deprecated.
```

File: trio/_threads.py
Comment (oremanj) [inline trio/_threads.py:203]: Please remove discussion of `cancellable` from the documentation (deprecated options don't get docs), and raise a deprecation warning if it is passed. You will need to default it to None in order to notice if someone passed False explicitly. You can make it a hard error to supply both `cancellable` and `abandon_on_cancel`, so you don't have to worry about which one overrides the other one.

**Comment by richardsheridan** [comment]
Guess I'm outvoted here. Hope everything goes ok downstream!

**Comment by richardsheridan** [comment]
> ```trio/_threads.py:295:9: B028 No explicit `stacklevel` keyword argument found```

I guess I don't know at what level of the stack to warn?

File: trio/_tests/test_threads.py
Comment (A5rocks) [inline trio/_tests/test_threads.py:1087]: The two `to_thread_run_sync`s should have a `# type: ignore[call-overload]`

File: trio/_tests/test_threads.py
Comment (A5rocks) [inline trio/_tests/test_threads.py:895]: Because `to_thread_run_sync` is an overloaded function, the error code associated with it changed from `arg-type` to `call-overload`.

This also applies to line 586.

**Review by A5rocks — COMMENTED** [review]
Re: mypy errors

File: trio/_threads.py
Comment (A5rocks) [inline trio/_threads.py:198]: For the two above I think you'll need to transfer over this `# type: ignore` over.

**Comment by oremanj** [comment]
> I'd sooner abandon this pull request than start raising warnings everywhere. The extra maintenance burden is 3 easy lines of code on our end, but a warning will send our users scrambling to perform busywork. Someone else should add a warning in a future PR for a real reason, like cancellable will be removed in 1.0, at the time such a decision is made to balance user and maintainer burdens.

Sorry, I reviewed without having seen this comment or I would've responded to it initially. I have empathy for this view, but it doesn't match how we've handled all other deprecations in the past, and I think the old name is actively harmful/confusing under the new semantics. (`cancellable=False` threads are IMO "more cancellable" than `cancellable=True` threads, because `cancellable=False` threads might actually terminate their operation if the `to_thread.run_sync` call is cancelled, while `cancellable=True` threads definitely won't.) I think most users of `to_thread.run_sync` probably don't specify `cancellable` at all, and for those that do, the warning is a nice prompt to reconsider whether they still need abandon-on-cancel now that actually-cancel-on-cancel is available.

File: trio/_threads.py
Comment (oremanj) [inline trio/_threads.py:295]: Suggest you use `trio._deprecate.warn_deprecated()` which provides a standard message format and warning class (so users can suppress Trio deprecation warnings independently of others).

File: trio/_threads.py
Comment (oremanj) [inline trio/_threads.py:296]:
```suggestion
            "The `cancellable=` keyword argument to `trio.to_thread.run_sync`",
```

File: trio/_threads.py
Comment (oremanj) [inline trio/_threads.py:299]:
```suggestion
            instead="`abandon_on_cancel=`",
```

Comment (oremanj) [reply trio/_threads.py:299]: These make the message be:
```
The `cancellable=` keyword argument to `trio.to_thread.run_sync` is deprecated since Trio 0.23.0; use `abandon_on_cancel=` instead (<issue url>)
```
which I think reads nicely.

**Review by oremanj — APPROVED** [review]

**Comment by A5rocks** [comment]
I'll merge this in soon-ish unless someone has other thoughts!

**Review by jakkdl — APPROVED** [review]
Sorry for being a pain in the butt, looks great!

**Comment by richardsheridan** [comment]
sorry thought the merge was a merge to master  not from master!

**Comment by jakkdl** [comment]
> sorry thought the merge was a merge to master not from master!

No worries! With the ongoing updates to linters etc I wanted to double check that this passes before merging even if GitHub deemed it a safe merge :)

**Pull request merged**
